tests/unit/s2n_client_hello_test.c

```diff
@@ -1397,6 +1397,175 @@ int main(int argc, char **argv)
+
+            EXPECT_SUCCESS(s2n_handshake_write_header(&client->handshake.io, TLS_CLIENT_HELLO));
+            EXPECT_SUCCESS(s2n_client_hello_send(client));
+            EXPECT_SUCCESS(s2n_handshake_finish_header(&client->handshake.io));
+
+            uint32_t raw_size = s2n_stuffer_data_available(&client->handshake.io);
+            EXPECT_NOT_EQUAL(raw_size, 0);
+            uint8_t *raw = s2n_stuffer_raw_read(&client->handshake.io, raw_size);
+            EXPECT_NOT_NULL(raw);
+
+            DEFER_CLEANUP(struct s2n_client_hello *client_hello = NULL, s2n_client_hello_free);
+            EXPECT_NOT_NULL(client_hello = s2n_client_hello_parse_message(raw, raw_size));
+            EXPECT_TRUE(client_hello->alloced);
+        };
+
+        /* Test: Rejects invalid ClientHellos
+         *
+         * This test is important to verify that no memory is leaked when parsing fails.
+         */
+        {
+            struct s2n_client_hello *client_hello = NULL;
+
+            uint8_t wrong_message_type[50] = { 0x02, 0x00, 0x00, 1 };
+            client_hello = s2n_client_hello_parse_message(wrong_message_type, sizeof(wrong_message_type));
+            EXPECT_NULL(client_hello);
+            EXPECT_EQUAL(s2n_errno, S2N_ERR_BAD_MESSAGE);
+
+            uint8_t wrong_message_size[50] = { 0x01, 0x00, 0x00, UINT8_MAX };
+            client_hello = s2n_client_hello_parse_message(wrong_message_size, sizeof(wrong_message_size));
+            EXPECT_NULL(client_hello);
+            EXPECT_EQUAL(s2n_errno, S2N_ERR_BAD_MESSAGE);
+
+            uint8_t too_short[5] = { 0x01, 0x00, 0x00, 1 };
+            client_hello = s2n_client_hello_parse_message(too_short, sizeof(too_short));
+            EXPECT_NULL(client_hello);
+            EXPECT_EQUAL(s2n_errno, S2N_ERR_STUFFER_OUT_OF_DATA);
+
+            uint8_t all_zeroes[50] = { 0x01, 0x00, 0x00, 46 };
+            client_hello = s2n_client_hello_parse_message(all_zeroes, sizeof(all_zeroes));
+            EXPECT_NULL(client_hello);
+            EXPECT_EQUAL(s2n_errno, S2N_ERR_BAD_MESSAGE);
+        };
+
+        /* Test: Rejects SSLv2 */
+        {
+            uint8_t sslv2_client_hello[] = {
+                SSLv2_CLIENT_HELLO_HEADER,
+                SSLv2_CLIENT_HELLO_PREFIX,
+                SSLv2_CLIENT_HELLO_CIPHER_SUITES,
+                SSLv2_CLIENT_HELLO_CHALLENGE,
+            };
+
+            /* Try parsing variations on the complete record vs just the message.
+             * The sslv2 record header is technically the first two bytes,
+             * but s2n-tls usually starts parsing after the first five bytes.
+             */
+            for (size_t i = 0; i <= S2N_TLS_RECORD_HEADER_LENGTH; i++) {
+                struct s2n_client_hello *client_hello = s2n_client_hello_parse_message(
+                        sslv2_client_hello + i, sizeof(sslv2_client_hello) - i);
+                EXPECT_NULL(client_hello);
+                EXPECT_EQUAL(s2n_errno, S2N_ERR_BAD_MESSAGE);
```

Comment:
It's a different format. s2n_client_hello_parse_raw doesn't check for it because that check is actually done earlier in our parsing. SSLv2 ClientHellos aren't standard handshake messages in standard records: when we try to read the record header, we also get part of the ClientHello. We've handled that by parsing some of the ClientHello when we normally parse the record header.

I could fix this by changing this old logic to just pass the whole ClientHello on, but that seems pointless since SSLv2 is EXTREMELY old and should absolutely not exist in the wild.